### PR TITLE
feat : add document errors, transition, validation

### PR DESCRIPTION
**Pull request by qrayven**
includes PR  #6 and #7

### CODE REVIEW

File: dpp/Cargo.toml
Comment (antouhou) [inline dpp/Cargo.toml]: No newline

File: dpp/src/data_contract/data_contract_factory.rs
Comment (antouhou) [inline dpp/src/data_contract/data_contract_factory.rs]: I probably wouldn't use dpp here and just pass the version from DPP constructor

Comment (antouhou) [inline dpp/src/data_contract/data_contract_factory.rs]: I probably wouldn't use dpp here and just pass the version from DPP constructor

Comment (antouhou) [reply dpp/src/data_contract/data_contract_factory.rs]: The reason is that later down the road passing dpp to factories will be quite problematic

File: dpp/src/data_contract/enrich_data_contract_with_base_schema.rs
Comment (antouhou) [inline dpp/src/data_contract/enrich_data_contract_with_base_schema.rs:20]: Does this actually work for `JsonValue`? Or is there an additional trait on json value that provides this?

Comment (antouhou) [inline dpp/src/data_contract/enrich_data_contract_with_base_schema.rs:20]: Does this actually work for `JsonValue`? Or is there an additional trait on json value that provides this?

File: dpp/src/data_contract/state_transition/data_contract_update_transition/validation/state/validate_data_contract_update_transition_state_factory.rs
Comment (antouhou) [inline dpp/src/data_contract/state_transition/data_contract_update_transition/validation/state/validate_data_contract_update_transition_state_factory.rs]: I would rename this to something else, as you're creating an instance of factory, which is not actually the case - you're creating the validator. I'd suggest naming it as something like `DataContractUpdateTransitionStateValidator`

File: dpp/src/data_contract/get_property_definition_by_path.rs
Comment (antouhou) [inline dpp/src/data_contract/get_property_definition_by_path.rs:11]: Oh, neat! Didn't know that this could work with WASM as well, I definitely need to transition other schemas I've made before to work like that

File: dpp/src/data_contract/state_transition/data_contract_create_transition/apply_data_contract_create_transition_factory.rs
Comment (antouhou) [inline dpp/src/data_contract/state_transition/data_contract_create_transition/apply_data_contract_create_transition_factory.rs]: The same with the other factories I've mentioned above

File: dpp/src/data_contract/state_transition/data_contract_create_transition/validation/state/validate_data_contract_create_transition_state_factory.rs
Comment (antouhou) [inline dpp/src/data_contract/state_transition/data_contract_create_transition/validation/state/validate_data_contract_create_transition_state_factory.rs:36]: I'm actually wondering how that's going to work with WASM - I don't have any experience with async and WASM, would be interesting to check it out

File: dpp/src/document/state_transition/documents_batch_transition/validation/state/fetch_documents_factory.rs
Comment (antouhou) [inline dpp/src/document/state_transition/documents_batch_transition/validation/state/fetch_documents_factory.rs]: I'd probably rename it to something like `DocumentsFetcher` or `DocumentsRepository`

File: dpp/src/util/json_schema.rs
Comment (antouhou) [inline dpp/src/util/json_schema.rs:6]: Oh, I see, so that's the `JsonValue` extension :)

File: dpp/src/util/json_value.rs
Comment (antouhou) [inline dpp/src/util/json_value.rs:27]: I've made a similar extension in my current PR, but for some different functionality. Will merge it with this one once we merge this PR

File: dpp/Cargo.toml
Comment (qrayven) [reply dpp/Cargo.toml]: fixed

File: dpp/src/data_contract/data_contract_factory.rs
Comment (qrayven) [reply dpp/src/data_contract/data_contract_factory.rs]: a really good point, fixed

Comment (qrayven) [reply dpp/src/data_contract/data_contract_factory.rs]: fixed

File: dpp/src/data_contract/enrich_data_contract_with_base_schema.rs
Comment (qrayven) [reply dpp/src/data_contract/enrich_data_contract_with_base_schema.rs:20]: In the `utils`I've created the trait `JsonSchemaExt` that extends `JsonValue` with methods that are useful when working with json schema.

File: dpp/src/data_contract/get_property_definition_by_path.rs
Comment (qrayven) [reply dpp/src/data_contract/get_property_definition_by_path.rs:11]: I was surprised as well actually. As in theory 'reads some files'.

File: dpp/src/util/json_value.rs
Comment (qrayven) [reply dpp/src/util/json_value.rs:27]: Great, we think the same :)

File: dpp/src/document/state_transition/documents_batch_transition/validation/state/fetch_documents_factory.rs
Comment (qrayven) [reply dpp/src/document/state_transition/documents_batch_transition/validation/state/fetch_documents_factory.rs]: fixed

File: dpp/src/data_contract/state_transition/data_contract_update_transition/validation/state/validate_data_contract_update_transition_state_factory.rs
Comment (qrayven) [reply dpp/src/data_contract/state_transition/data_contract_update_transition/validation/state/validate_data_contract_update_transition_state_factory.rs]: good point, fixed

File: dpp/src/data_contract/state_transition/data_contract_create_transition/validation/state/validate_data_contract_create_transition_state_factory.rs
Comment (qrayven) [reply dpp/src/data_contract/state_transition/data_contract_create_transition/validation/state/validate_data_contract_create_transition_state_factory.rs:36]: Not much expirence from me either. While implementing I've learnt that the `future` is being sent out to the js event loop, so in theory, its good to have async functions instead of synchronous ones. This will require closer look for sure.

File: dpp/src/data_contract/state_transition/data_contract_create_transition/apply_data_contract_create_transition_factory.rs
Comment (qrayven) [reply dpp/src/data_contract/state_transition/data_contract_create_transition/apply_data_contract_create_transition_factory.rs]: fixed

**Review by antouhou — APPROVED** [review]
Looks good 👍

**Pull request merged**
